### PR TITLE
Asset Viewer

### PR DESCRIPTION
**Pull request by german-e-mas**
This PR implements an Asset Viewer, which is in charge of rendering models and worlds without the need of a websocket connection.

The logic comes from the Gazebo App. These steps were originally performed by the [SDFViewerComponent](https://github.com/gazebo-web/gazebosim-app/blob/gzweb2/src/app/model/sdfviewer/sdfviewer.component.ts), but now they will be inside gzweb.

The `AssetViewer` can use its `renderFromFiles` method to receive an array of model or world files and render the SDF into a scene. Additional parameters can be set into its constructor, to determine for example if the model should be scaled or PBR materials should be used.

Aside from the `AssetViewer`, other small fixes were included:
- There is a new dependency to correctly parse SDF (XML) files into JSON.
- Pose parsing from SDF files was not working properly.
- Collada Loader had an incorrect up-axis parsing and rotation. The included changes were made to match the output of the Collada Loader with what we see in Gazebo.
- PBR materials were being parsed regardless of the `enablePBR` flag being set.
- Submeshes were incorrectly positioned due to an error using `Math.min` instead of `Math.max`.

---

## Notes:

1. This doesn't bump the version of gzweb. We can do so in a following PR. Gazebo App will require this in order to match its dependency.
2. Model Lightning doesn't match the previous gzweb version. You can see the difference in lightning in the following images. As this requires further investigation, we can address it in a following PR in order to reduce scope.

| gzweb v1 | gzweb v2 |
| --- | --- |
| staging-app gazebosim org_gmas_fuel_models_Urban%20Stairwell%20Platform%20Centered%20Lights | localhost_3000_gmas_fuel_models_Urban%20Stairwell%20Platform%20Centered%20Lights |

Can you ptal @nkoenig ?

### CODE REVIEW

File: src/AssetViewer.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+   * An optional Fuel Server Bearer Token in case the resource is private.
+   */
+  token?: string;
+}
+
+/**
+ * The Asset Viewer class allows clients to render and view simulation resources, such as
+ * models and worlds.
+ *
+ * This requires all of thhe resource's related URLs, and there is no websocket connection involved
```
Comment (nkoenig) [inline src/AssetViewer.ts:43]:
```suggestion
 * This requires all of the resource's related URLs, and there is no websocket connection involved
```

File: src/AssetViewer.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  public renderFromFiles(files: string[]): void {
+    if (!this.scene || !this.sdfParser) {
+      return;
+    }
+
+    this.sdfParser.usingFilesUrls = true;
+    this.sdfParser.enablePBR = this.shouldUsePBR;
+
+    // Look for SDF file.
+    const sdfFile = files.find((file) => file.endsWith(".sdf"));
```
Comment (nkoenig) [inline src/AssetViewer.ts:189]: This is a bit dangerous. A resource could contain multiple SDF files where only 1 should be rendered. Can you add a todo here:
```
TODO: Use the data within metadata.pbtxt to determine which SDF files to render.
```

File: src/AssetViewer.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+        this.scene?.add(obj);
+
+        this.resourceLoaded$.next(true);
+      });
+    }
+  }
+
+  /**
+   * Auxiliar method to scale the model. We aim to have it's largest dimension
```
Comment (nkoenig) [inline src/AssetViewer.ts:207]:
```suggestion
   * Auxiliary method to scale the model. We aim to have it's largest dimension
```

File: src/FuelServer.ts [hunk 4]
```diff
@@ -17,27 +17,28 @@ export const IGN_FUEL_HOST: string = 'fuel.ignitionrobotics.org';
...
-      return 'https://' + uriArray.join('/');
+  // Check if it's already a Fuel URI.
+  if (uri.startsWith(`https://${FUEL_HOST}`) || uri.startsWith(`https://${IGN_FUEL_HOST}`)) {
+    return uri;
+  }
+
+  // Check to see if the uri has the form similar to
+  // `/home/.../fuel.ignitionrobotics.org/...`
+  // If so, then we assume that the parts following
+  // `fuel.ignitionrobotics.org` can be directly mapped to a valid URL on
```
Comment (nkoenig) [inline src/FuelServer.ts:28]:
```suggestion
  // `fuel.gazebosim.org` can be directly mapped to a valid URL on
```

File: src/FuelServer.ts [hunk 5]
```diff
@@ -17,27 +17,28 @@ export const IGN_FUEL_HOST: string = 'fuel.ignitionrobotics.org';
...
-      uriArray.splice(1, 0, FUEL_VERSION);
-      uriArray.splice(6, 0, 'files');
-      return 'https://' + uriArray.join('/');
+  // Check if it's already a Fuel URI.
+  if (uri.startsWith(`https://${FUEL_HOST}`) || uri.startsWith(`https://${IGN_FUEL_HOST}`)) {
+    return uri;
+  }
+
+  // Check to see if the uri has the form similar to
+  // `/home/.../fuel.ignitionrobotics.org/...`
```
Comment (nkoenig) [inline src/FuelServer.ts:26]:
```suggestion
  // `/home/.../fuel.gazebosim.org/...`
```

File: src/AssetViewer.ts (hunk 2 above)
Comment (german-e-mas) [reply src/AssetViewer.ts:189]: Thank you! Yes, at first I was going to read `model.sdf` but there are many models and worlds that don't follow this structure.

Can we use `metadata.pbtxt` reliably as a source of truth? Is that file always available or required to upload a model or world? If not, what is a good way to determine what SDF to load?

File: include/ColladaLoader.js [hunk 6]
```diff
@@ -298,7 +304,9 @@ class ColladaLoader extends Loader {
 
 		function parseAssetUpAxis( xml ) {
 
-			return xml !== undefined ? xml.textContent : 'Y_UP';
+			// Modified by German Mas.
+			// Hardcode Y_UP in order to prevent unnecesary rotations on meshes.
+			return 'Y_UP';
```
[the remark this answers is not in the document]
Comment (german-e-mas) [reply include/ColladaLoader.js:309]: It does seem odd, I aggree. But if we don't do this, some meshes that have a certain up axis are not properly rendered.

There was a change on this logic when we migrated to Typescript, and it differs from what the previous version's Collada Loader had (see https://github.com/gazebo-web/gzweb/blob/main/include/ColladaLoader.js).

[the remark this answers is not in the document]
Comment (german-e-mas) [reply include/ColladaLoader.js:309]: Summary of the changes:
- gzweb2 rotates `Y_UP` by `(pi/2, 0, 0)`. No changes in the parsed the Up Axis.
- this branch hardcodes `Y_UP`. No additional rotations are applied. This is exactly the same we used for `gzweb 1`.
- Original ColladaLoader rotates `Z_UP` by `(-pi/2, 0, 0)`. No changes in the parsed the Up Axis.

With this into account, let's see some examples of problematic models. These use meshes with different up axes and they are being rendered differently, depending on the changes of the Collada Loader.

## marble_husky_sensor_config_1
- [Production](https://app.gazebosim.org/OpenRobotics/fuel/models/marble_husky_sensor_config_1)
- [Staging](https://staging-app.gazebosim.org/gmas/fuel/models/TestPose4)

This one has the following meshes:

| Mesh | Up Axis |
| --- | --- |
| VLP16_scan.dae | Y_UP |
| VLP16_base_2.dae | Y_UP |
| VLP16_base_1.dae | Y_UP |
| spotlight.dae | Z_UP |
| realsense.dae | Z_UP |
| base_link.dae | Z_UP |

And this is how it is being rendered (right-click, open image in a new tab):

| Gazebo | Gzweb2 Branch | This branch | Source THREE.ColladaLoader (no axis-related modifications) |
| --- | --- | --- | --- |
| Screenshot from 2023-04-24 14-23-58 | localhost_3000_gmas_fuel_models_TestPose4 (1) | localhost_3000_gmas_fuel_models_TestPose4 (3) | localhost_3000_gmas_fuel_models_TestPose5 (2) |

You can see that the VLP16 sensor is wrongly rotated in the gzweb2 branch. If we use the rotations of the original Collada Loader, anything but the sensor is wrongly rotated.

## CERBERUS_GAGARIN_SENSOR_CONFIG_1
- [Production](https://app.gazebosim.org/OpenRobotics/fuel/models/CERBERUS_GAGARIN_SENSOR_CONFIG_1)
- [Staging](https://staging-app.gazebosim.org/gmas/fuel/models/TestPose5)

The meshes are:

| Mesh | Up Axis |
| --- | --- |
| propeller_cw.dae | Y_UP |
| propeller_ccw.dae | Y_UP |
| cree_xhp70.dae | Z_UP |
| gagarin.dae | Z_UP |

Screenshots of each render:

| Gazebo | Gzweb2 Branch | This branch | Source THREE.ColladaLoader (no axis-related modifications) |
| --- | --- | --- | --- |
| Screenshot from 2023-04-24 13-50-49 | localhost_3000_gmas_fuel_models_TestPose5 (3) | localhost_3000_gmas_fuel_models_TestPose5 | localhost_3000_gmas_fuel_models_TestPose5 (1) |

You can see that the propellers are also wrongly rotated on gzweb2 branch, and the original Collada Loader rotates everything except for the propellers.

## Test case - barcs_qav500_no_sensors
- [Production](https://app.gazebosim.org/german/fuel/models/barcs_qav500_no_sensors)
- [Staging](https://staging-app.gazebosim.org/gmas/fuel/models/TestPose3)

This model was modified on purpose to see how these differences are applied.

Originally, [barcs_qav500_sensor_config_1](https://app.gazebosim.org/OpenRobotics/fuel/models/barcs_qav500_sensor_config_1) is very similar to the CERBERUS_GAGARIN_SENSOR_CONFIG_1 case. Its propellers are Y_UP and they had an incorrect rotation. **For this experiment, I changed one of the propeller's pair to Z_UP instead.**

| Mesh | Up Axis |
| --- | --- |
| neo11_propeller_cw.dae | Y_UP |
| neo11_propeller_ccw.dae | Z_UP\* |
| spotlight.dae | Z_UP |


And this is how it is being rendered. Important: The UAV's body is a STL mesh, so it doesn't go through the Collada Loader but the STL one.

| Gazebo | Gzweb2 Branch | This branch | Source THREE.ColladaLoader (no axis-related modifications) |
| --- | --- | --- | --- |
| Screenshot from 2023-04-24 14-59-44 | localhost_3000_gmas_fuel_models_TestPose4 (2) | localhost_3000_gmas_fuel_models_TestPose3 (2) | localhost_3000_gmas_fuel_models_TestPose3 (3) |

Note that Gazebo does not make any difference between the two propellers, while they are rotated in the gzweb2 branch.

---

The current changes (hardcoding Y_UP and not dealing with rotations) maintains the behavior seen in Gazebo.

Let me know what do you think @nkoenig

File: include/ColladaLoader.js [hunk 7]
```diff
@@ -4280,9 +4288,9 @@ class ColladaLoader extends Loader {
 		const scene = parseScene( getElementsByTagName( collada, 'scene' )[ 0 ] );
 		scene.animations = animations;
 
-		if ( asset.upAxis === 'Y_UP' ) {
+		if ( asset.upAxis === 'Z_UP' ) {
```
[the remark this answers is not in the document]
Comment (german-e-mas) [reply include/ColladaLoader.js:4291]: See my comment on https://github.com/gazebo-web/gzweb/pull/50/files#r1175656747

File: include/ColladaLoader.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (nkoenig) [reply include/ColladaLoader.js:309]: Thanks, looks good.

**Review by nkoenig — APPROVED** [review]

**Comment by nkoenig** [comment]
@german-e-mas , is this safe to merge?

**Comment by german-e-mas** [comment]
> @german-e-mas , is this safe to merge?

It is. I will double check with the Gazebo App changes.

**Comment by german-e-mas** [comment]
Merging this @nkoenig . I'll open a new PR with the color space changes, to see models with the same colors as before.

**Pull request merged**